[5.2] [Guided tours] Auto start tours - full functionality

obuisard opened the pull request:
### Summary of Changes

The welcome tour has been introduced in Joomla 5.1. It is a tour that starts automatically any time a new instance of Joomla has been installed. However, at that time, we could not introduce a full implementation. Therefore, the welcome tour was meant to run only once and pretty much only for the person who installed Joomla.

This PR is meant to add the missing functionality of the auto-starting tours.

It adds several enhancements:
- a way for the users to opt-out of this functionality, as auto-starting tours can be a 'nuisance' for some (although tours can quickly be discarded by selecting the new button 'Hide Forever'),
- a way to log user behavior in the User Action Logs (by opting-in the guided tours in the action logs, one can find out if a user skipped, completed or cancelled a tour and at what step) - which works for all tour, not just the auto-starting ones,
- a way to make any tour an auto-starting one (see complementary PR #43690),
- a way to store the tour state for every user in the User Profiles table (specific to auto-starting tours),
- the possibility to run a tour after a specific delay when a user has cancelled an auto-started tour,
- the possibility to keep running tours manually even if they have been skipped or previously completed.

Many thanks to Shirielle Williams @sdwjoomla, Marc Studer (co-implementer), Adam Melcher and the Guided Tours team for their work and feedback.

### Testing Instructions

Download the full and update packages from the [Downloads ](https://github.com/joomla/joomla-cms/pull/43814#partial-pull-merging) checks. 
You cannot test this PR through the patch tester extension.

Test 1 - on a new instance
----------------------------
Install the full package of this PR.
The welcome tour will run and you should see a button 'Hide Forever'. Do not select the button.
Cancel the tour (with the X).

Go to Guided Tour's global configuration.
Set the time for the delay parameter to 1 minute (this will help testing the cancelling of tours when they auto-start).
Save.
When a tour auto-starts, cancelling it prevents the tour from running again for a certain period of time (hence the delay parameter).

Go to Users -> Users Action Logs
Filter by the 'Guided Tours' extension. The extension will NOT be in the list if there are no recorded logs yet.
You should see User [you] delayed the tour 'Welcome to Joomla!'.
On a new instance of Joomla, the Guided Tours action logs are enabled by default.
It has probably been a minute since you cancelled the tour, so go back to the dashboard.
The welcome tour should run again.
Cancel the tour (with the X).

Test 2 - opt-out of auto-starting tours
Goal: disable auto-starting tours for all users 
------------------------------------------
Go to Guided Tour's global configuration.
Disallow the auto-starting of the tours. 
Save.
Go back to the dashboard.
Make sure a minute has passed.
Refresh the page if necessary.
The welcome tour should no longer start.

Test 3 - opt-out of auto-starting tours for a single user
Goal: disable auto-starting tours for one user
------------------------------------------------------------
Go to Guided Tour's global configuration.
Allow the auto-starting of the tours. 
Save.
Go back to the dashboard.
The welcome tour should run again.
Cancel the tour (with the X).
Go to your user profile.
In the 'Basic settings' tab, select the option 'no' for the 'auto starting tours' parameter.
Save.
Go back to the dashboard.
Make sure a minute has passed.
Refresh the page if necessary.
The welcome tour should no longer start.

Test 4 - opt-out of the welcome tour
Goal: opt-out of a tour that starts automatically
----------------------------------------
Go to your user profile.
In the 'Basic settings' tab, select the option 'Use default' for the 'auto starting tours' parameter.
Save.
Go back to the dashboard.
The welcome tour should run again.
Select 'Hide Forever' button to dismiss the tour completely.
Go to Users -> Users Action Logs
Filter by the 'Guided Tours' extension.
You should see User [you] skipped the tour 'Welcome to Joomla!'.
Go back to the dashboard.
Make sure a minute has passed.
The tour should no longer start by itself.

Test 5 - run the welcome tour manually
Goal: possibility to run tours that have been skipped previously
-------------------------------------------
Go to the guided tour list of tours at the top of the dashboard and run the welcome tour.
The 'Hide forever' button should be missing.
Running the tour manually should not interfere with your decision to skip the tour or if you have completed it. In those 2 cases, the tour should never auto-start again.

Test 6 - auto start 2 welcome tours
Goal: test the sequencing of tours when several tours are set to run on the same page
-------------------------------------
Go to System -> Guided Tours (in the 'Manage' section).
Select the welcome tour (it should be the last on the list of tours).
Save as copy. Publish the tour. Change the title to your liking (by replacing the language key). Save and close.
No need to add steps for this test.
Copy the tour a second time (change title, publish and save) to create a 3rd welcome tour (remember, the first welcome tour should never show again).
Now you have 2 auto-starting tours ready to go.
Go to the home dashboard.
The first auto-starting tour should start. Cancel by clicking on the X.
Nothing else should happen.
Go to Content -> Articles (or just refresh the page). Go back to the home dashboard.
The second tour should start unless the first tour's delay has already passed.
Cancel the tour with 'X'.
Wait one minute, refresh the page. The first tour should start again. Hide forever.
Wait one minute, refresh the page. The second tour should start. Hide forever.
No tour should show again after one minute, no matter what you do (refresh the page, do something - like checking the user action logs - and go back to the home dashboard).

Test 7 - on an updated instance
Goal: test settings on update
-----------------------------------
Install the update package of this PR over a 5.1 instance or earlier version of 5.2 (there was no welcome tour in 5.0).
The welcome tour should not run, it has been disabled from auto-starting after it has ran once in Joomla 5.1. 
Go to the User Action Logs global configuration.
The Guided Tours should not be selected.

Note
-----

PR #43690 has been merged and therefore you can now change auto-start settings for tours when editing a tour.

![image](https://github.com/user-attachments/assets/3c5949cf-0b0d-4c40-a798-1e2dee44a6c8)

### Actual result BEFORE applying this Pull Request

The welcome tour starts and only one user can see it. It runs only once. To see it again, the tour needs to be run manually.

### Expected result AFTER applying this Pull Request

Any tour set to auto-start should automatically run, one at a time, when in the context it is supposed to run under.
Tours run automatically for all users logging in the backend (if the tour is available for the user group the user is part of).
When a tour is cancelled, it restarts after a certain delay. When skipped (hidden forever) or completed, the tour no longer runs.
All tours are still available to run manually.
Manual tours do not have the button 'Hide forever'.

### Link to documentations
Please select:
- [x] Documentation link for docs.joomla.org: [Guided Tours Options](https://docs.joomla.org/Help5.x:Guided_Tours:_Options)
- [ ] No documentation changes for docs.joomla.org needed

- [x] Pull Request link for manual.joomla.org: [Auto-start and events](https://github.com/joomla/Manual/pull/290)
- [ ] No documentation changes for manual.joomla.org needed

Note:
User and User Action Logs help screens are severally out of date and it's impossible to add information related to this PR without adding the already missing information and re-creating screenshots.

obuisard committed:
Added translations auto start parameter

obuisard committed:
Added translations auto-start parameter

obuisard committed:
Added tourAutoStart parameter to allow a user to opt out of auto-starting tours

obuisard committed:
Added TourAutoStart parameter to allow users to opt out of auto-starting tours.

obuisard committed:
Added parameters to opt out of auto-starting tours and to set the delay after which the tour will run again

obuisard committed:
Added translations handling the new parameters for config.xml

obuisard committed:
Create 5.2.0-2024-07-19.sql to add Guided Tours to the list of log able actions

obuisard committed:
Create 5.2.0-2024-07-19.sql to add Guided Tours to the list of log able actions

obuisard committed:
Added onTourRunSaveState to log user behavior while running tours

obuisard committed:
Added translations to cover the messages handling user tour run state

obuisard committed:
Update Joomla.php

Added spaces

obuisard committed:
Create AjaxController.php

Adds a way to store user state information when a user skip, complete or cancel a tour

obuisard committed:
Update AjaxController.php

Renamed params for onTourRunSaveState

obuisard committed:
Changes in the naming/order of variables

obuisard committed:
Added user state translations

obuisard committed:
Changed language key name

obuisard committed:
Added language keys and translations to skip tours and what's new tours

obuisard committed:
Added functionality that supports auto-starting the tours

obuisard committed:
Adde functionality for auto-starting tours

obuisard committed:
Added error message for the Ajax calls

obuisard committed:
Removed tabs

obuisard committed:
Removed tabs

obuisard committed:
Added Guided Tours as default loggable extension

obuisard committed:
Removed/added spaces

obuisard committed:
Removed/added spaces

obuisard committed:
Removed trailing spaces.

obuisard committed:
Added space after switch keyword

obuisard committed:
Update guidedtours.es6.js

Many structural changes, quotes, tab fixes

obuisard committed:
Removed unnecessary variables

obuisard committed:
Update provider.php

Added database support

obuisard committed:
Added guided tours to action logs extensions

obuisard committed:
Added guided tours to action logs extensions

obuisard committed:
Changed setval

brianteeman commented on administrator/components/com_guidedtours/config.xml:
```diff
@@ -1,5 +1,28 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <config>
+	<fieldset name="guidedtours_config" label="COM_GUIDEDTOURS">
+		<field
+			name="tourAutoStart"
+			type="radio"
+			layout="joomla.form.field.radio.switcher"
+			label="COM_GUIDEDTOURS_CONFIG_USER_TOURAUTOSTART_LABEL"
+			description="COM_GUIDEDTOURS_CONFIG_USER_TOURAUTOSTART_DESCRIPTION"
+			default="allow"
```
please reorder name, type, label, description ... to follow our code style

obuisard replied:
Definitely needs a cleanup in the core. For instance, com_content and com_contact don't follow the same code style (layout before or after the default for instance). com_banners has layout before label in some cases...
I put the 'layout' before 'default'. Not sure it's alright then...

brianteeman replied:
the only thing that is in the codestyle is name, type, label, description. After that there is nothing stipulated https://developer.joomla.org/coding-standards/xml.html

 If there is anything different anywhere in core for those - then its a bug.

obuisard replied:
Thanks Brian.

brianteeman commented on administrator/language/en-GB/com_guidedtours.ini:
```diff
@@ -4,6 +4,12 @@
 COM_GUIDEDTOURS="Guided Tours"
+COM_GUIDEDTOURS_CONFIG_DELAYED_TIME_DESCRIPTION="The amount of time (in seconds) a tour is delayed after being cancelled by the user and until it is shown again (only when a tour is set to start automatically).<br />For instance, enter 3600 for 1 hour, 86400 for 24 hours, 604800 for 1 week."
+COM_GUIDEDTOURS_CONFIG_DELAYED_TIME_LABEL="Auto Start Time Delay"
+COM_GUIDEDTOURS_CONFIG_USER_TOURAUTOSTART_DESCRIPTION="Turn on or off the auto starting functionality of tours"
```
```suggestion
COM_GUIDEDTOURS_CONFIG_USER_TOURAUTOSTART_DESCRIPTION="Turn on or off the auto starting functionality of tours."
```

A sentence should end with a .

brianteeman commented on administrator/language/en-GB/com_guidedtours.ini:
```diff
@@ -88,4 +94,7 @@ COM_GUIDEDTOURS_TYPE_REDIRECT_URL_DESC="Enter the relative URL of the page you w
 COM_GUIDEDTOURS_URL_LABEL="Relative URL"
 COM_GUIDEDTOURS_URL_DESC="Enter the relative URL of the page from where you want to Start the tour, e.g administrator/index.php?option=com_guidedtours&view=tours for the tours' list page."
+COM_GUIDEDTOURS_USERSTATE_CONNECTEDONLY="Tour User state action is only for connected users."
+COM_GUIDEDTOURS_USERSTATE_STATENOTSAVED="User state not saved for user %s tour %s."
```
```suggestion
COM_GUIDEDTOURS_USERSTATE_STATENOTSAVED="User state not saved for user %1$s tour %2$s."
```

https://docs.joomla.org/Formatted_fields_in_language_translation_strings

**Format Argument Swapping**
The format string supports argument numbering and even swapping. This is useful where two or more data items must be embedded in a string but differences in language structure means that the order of use of the data items is not the same.

obuisard replied:
Thanks Brian

brianteeman commented on administrator/language/en-GB/com_guidedtours.ini:
```diff
@@ -88,4 +94,7 @@ COM_GUIDEDTOURS_TYPE_REDIRECT_URL_DESC="Enter the relative URL of the page you w
 COM_GUIDEDTOURS_URL_DESC="Enter the relative URL of the page from where you want to Start the tour, e.g administrator/index.php?option=com_guidedtours&view=tours for the tours' list page."
+COM_GUIDEDTOURS_USERSTATE_CONNECTEDONLY="Tour User state action is only for connected users."
+COM_GUIDEDTOURS_USERSTATE_STATENOTSAVED="User state not saved for user %s tour %s."
+COM_GUIDEDTOURS_USERSTATE_STATESAVED="User state saved for user %s tour %s."
```
```suggestion
COM_GUIDEDTOURS_USERSTATE_STATESAVED="User state saved for user %1$s tour %2$s."
```


https://docs.joomla.org/Formatted_fields_in_language_translation_strings

**Format Argument Swapping**
The format string supports argument numbering and even swapping. This is useful where two or more data items must be embedded in a string but differences in language structure means that the order of use of the data items is not the same.

brianteeman commented on build/media_source/plg_system_guidedtours/js/guidedtours.es6.js:
```diff
@@ -15,6 +15,71 @@ function emptyStorage() {
+}
+
+/**
+  Synchronize tour state for this user in his account/profile
```
```suggestion
  Synchronize tour state for this user in their account/profile
```

please try to use gender neutral labuage

obuisard committed:
Update config.xml

Moved layout attribute.

obuisard committed:
Update administrator/language/en-GB/com_guidedtours.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

obuisard committed:
Update administrator/language/en-GB/com_guidedtours.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

obuisard committed:
Update administrator/language/en-GB/com_guidedtours.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

obuisard committed:
Update build/media_source/plg_system_guidedtours/js/guidedtours.es6.js

Co-authored-by: Brian Teeman <brian@teeman.net>

obuisard committed:
Swapped cancelTour with skipTour so that the tour does not run again

Quy commented on administrator/components/com_guidedtours/src/Controller/AjaxController.php:
```diff
@@ -0,0 +1,180 @@
+     */
+    public function fetchUserState()
+    {
+        $user   = $this->app->getIdentity();
+
+        $tourId     = $this->app->input->getInt('tid', 0);
+        $stepNumber = $this->app->input->getString('sid', '');
+        $context    = $this->app->input->getString('context', '');
+
+        if ($user != null && $user->id > 0) {
+            $actionState   = '';
+
+            switch ($context) {
+                case "tour.complete":
+                    $actionState = "completed";
+                    break;
+                case "tour.cancel":
+                    $actionState = "delayed";
+                    break;
+                case "tour.skip":
+                    $actionState = "skipped";
+                    break;
+            }
```
```suggestion
        $user = $this->app->getIdentity();

        $tourId     = $this->app->input->getInt('tid', 0);
        $stepNumber = $this->app->input->getString('sid', '');
        $context    = $this->app->input->getString('context', '');

        if ($user != null && $user->id > 0) {
            $actionState = '';

            switch ($context) {
                case 'tour.complete':
                    $actionState = 'completed';
                    break;
                case 'tour.cancel':
                    $actionState = 'delayed';
                    break;
                case 'tour.skip':
                    $actionState = 'skipped';
                    break;
            }
```

obuisard replied:
Thanks!

obuisard committed:
Update administrator/components/com_guidedtours/src/Controller/AjaxController.php

Co-authored-by: Quy <quy@nomonkeybiz.com>

obuisard committed:
Increased delay to 1 hour

obuisard committed:
Default delay to one hour

obuisard committed:
Merge branch '5.2-dev' into guided-tours-autostart

Quy commented on administrator/language/en-GB/com_guidedtours.ini:
```diff
@@ -6,6 +6,12 @@
 COM_GUIDEDTOURS_AUTOSTART_DESC="Start the tour automatically when a user reaches the context in which the tour should be displayed."
 COM_GUIDEDTOURS_AUTOSTART_LABEL="Auto Start"
+COM_GUIDEDTOURS_CONFIG_DELAYED_TIME_DESCRIPTION="The amount of time (in seconds) a tour is delayed after being cancelled by the user and until it is shown again (only when a tour is set to start automatically).<br />For instance, enter 3600 for 1 hour, 86400 for 24 hours, 604800 for 1 week."
+COM_GUIDEDTOURS_CONFIG_DELAYED_TIME_LABEL="Auto Start Time Delay"
```
```suggestion
COM_GUIDEDTOURS_CONFIG_DELAYED_TIME_LABEL="Auto Start Time Delay (in seconds)"
```

obuisard replied:
Thanks :-)

Quy commented on administrator/components/com_guidedtours/config.xml:
```diff
@@ -1,5 +1,28 @@
+			layout="joomla.form.field.radio.switcher"
+			default="allow"
+			>
+			<option value="disallow">COM_GUIDEDTOURS_CONFIG_USER_TOURAUTOSTART_OPTION_DISALLOW</option>
+			<option value="allow">COM_GUIDEDTOURS_CONFIG_USER_TOURAUTOSTART_OPTION_ALLOW</option>
```
Use `Yes` or `No`?

obuisard replied:
I agree, it would be better, less keys to translate, more concise... I modified the code to reflect this

Quy commented on administrator/components/com_guidedtours/src/Controller/AjaxController.php:
```diff
@@ -0,0 +1,180 @@
+            if (!\is_null($result)) {
+                $values = json_decode($result, true);
+
+                if (!empty($values)) {
+                    // The profile is updated only when delayed. 'Completed' and 'Skipped' are final
+                    if ($values['state'] === 'delayed') {
+                        $db->updateObject('#__user_profiles', $profileObject, ['user_id', 'profile_key']);
+                    }
+                }
```
```suggestion
                // The profile is updated only when delayed. 'Completed' and 'Skipped' are final
                if (!empty($values) && $values['state'] === 'delayed') {
                    $db->updateObject('#__user_profiles', $profileObject, ['user_id', 'profile_key']);
                }
```

obuisard replied:
Thanks!

obuisard committed:
Update administrator/language/en-GB/com_guidedtours.ini

Co-authored-by: Quy <quy@nomonkeybiz.com>

obuisard committed:
Update administrator/components/com_guidedtours/src/Controller/AjaxController.php

Co-authored-by: Quy <quy@nomonkeybiz.com>

obuisard committed:
Prefer yes/no to allow/disallow

obuisard committed:
Prefer yes/no to allow/disallow

obuisard committed:
Removed allow/disallow

obuisard committed:
Removed allow/disallow

obuisard committed:
Correctly ordered string

obuisard committed:
Use yes/no instead of allow/disallow

obuisard committed:
Removed allow/disallow

obuisard committed:
Parameter allowTourAutoStart is tested with yes/no values, no longer allow/disallow

obuisard committed:
Merge branch '5.2-dev' into guided-tours-autostart

Quy commented on plugins/system/guidedtours/src/Extension/GuidedTours.php:
```diff
@@ -141,41 +148,104 @@ public function onBeforeCompileHead()
+                        if (empty($values)) {
+                            // Do not start the tour.
+                            $canAutostart = false;
+                        } else {
+                            if ($values['state'] === 'skipped' || $values['state'] === 'completed') {
+                                $canAutostart = false;
+                            }
+                            if ($values['state'] === 'delayed') {
+                                $delay       = $params->get('delayed_time', '3600');
+                                $currentTime = Date::getInstance();
+                                $loggedTime  = new Date($values['time']['date']);
+
+                                if ($loggedTime->add(new \DateInterval('PT' . $delay . 'S')) > $currentTime) {
+                                    $canAutostart = false;
+                                }
+                            }
+                        }
```
```suggestion
                        } else if ($values['state'] === 'skipped' || $values['state'] === 'completed') {
                            $canAutostart = false;
                        } else if ($values['state'] === 'delayed') {
                            $delay       = $params->get('delayed_time', '3600');
                            $currentTime = Date::getInstance();
                            $loggedTime  = new Date($values['time']['date']);

                            if ($loggedTime->add(new \DateInterval('PT' . $delay . 'S')) > $currentTime) {
                                $canAutostart = false;
                            }
                        }
```

brianteeman commented on administrator/language/en-GB/com_guidedtours.ini:
```diff
@@ -6,6 +6,10 @@
 COM_GUIDEDTOURS="Guided Tours"
 COM_GUIDEDTOURS_AUTOSTART_DESC="Start the tour automatically when a user reaches the context in which the tour should be displayed."
 COM_GUIDEDTOURS_AUTOSTART_LABEL="Auto Start"
+COM_GUIDEDTOURS_CONFIG_DELAYED_TIME_DESCRIPTION="The amount of time (in seconds) a tour is delayed after being cancelled by the user and until it is shown again (only when a tour is set to start automatically).<br />For instance, enter 3600 for 1 hour, 86400 for 24 hours, 604800 for 1 week."
```
is seconds the most appropriate unit of time for this

obuisard replied:
I have made the appropriate changes to use minutes rather than seconds (config.xml, translations and testing in the plugin). Thank you.

brianteeman commented on administrator/language/en-GB/com_guidedtours.ini:
```diff
@@ -6,6 +6,10 @@
 COM_GUIDEDTOURS="Guided Tours"
 COM_GUIDEDTOURS_AUTOSTART_DESC="Start the tour automatically when a user reaches the context in which the tour should be displayed."
 COM_GUIDEDTOURS_AUTOSTART_LABEL="Auto Start"
+COM_GUIDEDTOURS_CONFIG_DELAYED_TIME_DESCRIPTION="The amount of time (in seconds) a tour is delayed after being cancelled by the user and until it is shown again (only when a tour is set to start automatically).<br />For instance, enter 3600 for 1 hour, 86400 for 24 hours, 604800 for 1 week."
```
```suggestion
COM_GUIDEDTOURS_CONFIG_DELAYED_TIME_DESCRIPTION="The amount of time (in seconds) a tour is delayed after being cancelled by the user and until it is shown again (only when a tour is set to start automatically).<br>For instance, enter 3600 for 1 hour, 86400 for 24 hours, 604800 for 1 week."
```

dont use xhtml formatting

obuisard replied:
All right, thanks Brian.

obuisard committed:
Update plugins/system/guidedtours/src/Extension/GuidedTours.php

Co-authored-by: Quy <quy@nomonkeybiz.com>

brianteeman commented on administrator/language/en-GB/com_guidedtours.ini:
```diff
@@ -90,4 +94,7 @@ COM_GUIDEDTOURS_TYPE_REDIRECT_URL_DESC="Enter the relative URL of the page you w
 COM_GUIDEDTOURS_URL_LABEL="Relative URL"
 COM_GUIDEDTOURS_URL_DESC="Enter the relative URL of the page from where you want to Start the tour, e.g administrator/index.php?option=com_guidedtours&view=tours for the tours' list page."
+COM_GUIDEDTOURS_USERSTATE_CONNECTEDONLY="Tour User state action is only for connected users."
+COM_GUIDEDTOURS_USERSTATE_STATENOTSAVED="User state not saved for user %1$s tour %2$s."
+COM_GUIDEDTOURS_USERSTATE_STATESAVED="User state saved for user %1$s tour %2$s."
```
should these not be "Tour User state ..

obuisard committed:
Update administrator/language/en-GB/com_guidedtours.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented on plugins/actionlog/joomla/src/Extension/Joomla.php:
```diff
@@ -1298,4 +1298,56 @@ public function onUserBeforeSave(User\BeforeSaveEvent $event): void
     }
+
+    /**
+     * Method is called when a user cancel, complete or skip a tour
```
```suggestion
     * Method is called when a user cancels, completes or skips a tour
```

brianteeman commented on plugins/system/guidedtours/src/Extension/GuidedTours.php:
```diff
@@ -141,41 +148,104 @@ public function onBeforeCompileHead()
+                return;
+            }
+
+            // The following code only relate to the auto-start functionality.
```
```suggestion
            // The following code only relates to the auto-start functionality.
```

obuisard committed:
Update plugins/actionlog/joomla/src/Extension/Joomla.php

Co-authored-by: Brian Teeman <brian@teeman.net>

obuisard committed:
Update plugins/system/guidedtours/src/Extension/GuidedTours.php

Co-authored-by: Brian Teeman <brian@teeman.net>

obuisard committed:
Changed wording for the tour user state

Quy commented on plugins/system/guidedtours/src/Extension/GuidedTours.php:
```diff
@@ -141,41 +148,101 @@ public function onBeforeCompileHead()
+                        if (empty($values)) {
+                            // Do not start the tour.
+                            $canAutostart = false;
+                        } else if ($values['state'] === 'skipped' || $values['state'] === 'completed') {
+                            $canAutostart = false;
+                        } else if ($values['state'] === 'delayed') {
```
```suggestion
                        } elseif ($values['state'] === 'skipped' || $values['state'] === 'completed') {
                            $canAutostart = false;
                        } elseif ($values['state'] === 'delayed') {
```

QuyTon committed:
cs

Quy commented:
> On a new instance of Joomla, the Guided Tours action logs are enabled by default.

This is not the case.
![43814-test1](https://github.com/user-attachments/assets/4b1a6444-8eaf-4d87-832d-b9b36c8ebc38)

obuisard commented:
> > On a new instance of Joomla, the Guided Tours action logs are enabled by default.

I forgot to add a database entry into #__action_log_config.

obuisard committed:
Added action_log_config

obuisard committed:
Added action_log_config

obuisard committed:
Added guided tours to action_log_config

obuisard committed:
Added guided tours to action_log_config

obuisard committed:
Replaced getCmd with get

obuisard committed:
Reflect changes made to Action Logs (new events)

obuisard committed:
Use AbstractEvent for onTourRunSaveState

obuisard committed:
Missing error messages

Quy commented on plugins/system/guidedtours/src/Extension/GuidedTours.php:
```diff
@@ -141,41 +148,103 @@ public function onBeforeCompileHead()
+
+                        if (empty($values)) {
+                            // Do not start the tour.
+                            $canAutostart = false;
```
Use `continue` rather than assign to a variable? Then no need to check variable in line 244.

obuisard replied:
Agree, it can be simpler.

Quy commented:
> I forgot to add a database entry into #__action_log_config.

It is still not enabled by default with the latest commits per the built-in package.

obuisard committed:
Restructured before and after events

obuisard committed:
Add guided tours to action logs

obuisard committed:
Added guided tours to action logs

obuisard commented:
> It is still not enabled by default with the latest commits per the built-in package.

It is fixed, I needed to add com_guidedtours to base.sql in the installation folder.

terryburg commented:
I have tested this item :red_circle: unsuccessfully on dfd42fc4e28b547daea7b71ab70ee9e5e2a04e68<br /><br />I used the full install package.

Not truly an unsuccessful test because almost everything major worked like they were supposed to, but some minor things didn't work like the directions said.

Test 1
"Go to Users -> Users Action Logs
Filter by the 'Guided Tours' extension." 
- ** there was no such filter option. That I could find anyhow. Only available were Users and Configuration. Anything I did to Tours got marked as extension Configuration in the logs, also. **

Test 4
"Go to Users -> Users Action Logs
Filter by the 'Guided Tours' extension." 
– **again no such filter option**
"You should see User [you] skipped the tour 'Welcome to Joomla!'." 
– ** did not see this, only that I updated the user, and prior to that that I had changed settings of the application configuration **.

Test 5
"If the tour is not completed and it had been cancelled before (neither skipped nor completed), it will auto-start after the delay set in the global configuration."
- ** I must have misunderstood these directions. Interrupted the manually started tour after two steps with the X option. The interrupted tour did not ever autorestart after that, and I am unclear why it should have at all, since it had been "hidden forever" in Step 4 **. 

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43814">issues.joomla.org/tracker/joomla-cms/43814</a>.</sub>

obuisard commented:
Thanks you for testing @terryburg !

> I have tested this item 🔴 unsuccessfully on [dfd42fc](https://github.com/joomla/joomla-cms/commit/dfd42fc4e28b547daea7b71ab70ee9e5e2a04e68)I used the full install package.

I am afraid this is not the latest version available. Major changes have been made after I discovered changes to the User Action Logs plugin today.

> Not truly an unsuccessful test because almost everything major worked like they were supposed to, but some minor things didn't work like the directions said.
> 
> Test 1 "Go to Users -> Users Action Logs Filter by the 'Guided Tours' extension."
> * ** there was no such filter option. That I could find anyhow. Only available were Users and Configuration. Anything I did to Tours got marked as extension Configuration in the logs, also. **

There is no 'Guided Tours' selection in the filter until there are recorded user action logs for it (I have now added such a comment in the test instructions).
You did not have the option to select 'Guided Tours', therefore there was no action log recorded.
Action logs were broken until I made corrections to the code to fix it.

> Test 4 "Go to Users -> Users Action Logs Filter by the 'Guided Tours' extension." – **again no such filter option** "You should see User [you] skipped the tour 'Welcome to Joomla!'." – ** did not see this, only that I updated the user, and prior to that that I had changed settings of the application configuration **.
> 

Same thing, action logs were broken therefore no user action log about guided tours has been recorded.

> Test 5 "If the tour is not completed and it had been cancelled before (neither skipped nor completed), it will auto-start after the delay set in the global configuration."
> 
> * ** I must have misunderstood these directions. Interrupted the manually started tour after two steps with the X option. The interrupted tour did not ever autorestart after that, and I am unclear why it should have at all, since it had been "hidden forever" in Step 4 **.

Yes, my explanations were not needed here, I am going to fix my comment.

obuisard committed:
Missing spaces

obuisard committed:
Update GuidedTours.php

Removed useless variable, use continue instead.

obuisard committed:
Changed return value of event

Quy commented on plugins/actionlog/joomla/src/Extension/Joomla.php:
```diff
@@ -1298,4 +1300,56 @@ public function onUserBeforeSave(User\BeforeSaveEvent $event): void
+     *
+     * @param   AbstractEvent $event The event instance.
+     *
+     * @return  boolean
```
```suggestion
     * @return  void
```

obuisard committed:
Update plugins/actionlog/joomla/src/Extension/Joomla.php

Co-authored-by: Quy <quy@nomonkeybiz.com>

Quy commented on plugins/system/guidedtours/src/Extension/GuidedTours.php:
```diff
@@ -259,6 +325,13 @@ private function processTour($item)
+        // What's new tours have a different label.
+        $temp->start_label = Text::_('PLG_SYSTEM_GUIDEDTOURS_START');
+        if (str_contains($item->uid, 'joomla-whatsnew')) {
+            $temp->start_label = Text::_('PLG_SYSTEM_GUIDEDTOURS_LEARNMORE');
```
Is this for another PR?

obuisard replied:
Yes, indeed.
A PR is on the way that will add a 'what's new' auto-starting tour to 5.2. Every minor version will get a new 'what's new tour' and every time the tour UID will contain 'joomla-whatsnew'. 
From now on, new installs will get a 'welcome' tour, updates will get a 'what's new tour.

brianteeman replied:
please dont mix pull requests - it makes it very hjard to test

obuisard replied:
Right, I will add the test in the what's new tour PR

brianteeman replied:
thank you

obuisard committed:
Removed code related to the what's new tour PR

obuisard committed:
Changed delay default value

obuisard committed:
Changed delay translation to reflect delays in minutes, not seconds

obuisard committed:
Delay is calculated in minutes, not seconds

obuisard committed:
Added comments when tours autostart

Quy commented on plugins/system/guidedtours/src/Extension/GuidedTours.php:
```diff
@@ -141,41 +148,102 @@ public function onBeforeCompileHead()
+                    if (!\is_null($result)) {
+                        $values = json_decode($result, true);
+
+                        if (empty($values)) {
+                            // Do not start the tour.
+                            continue;
+                        } elseif ($values['state'] === 'skipped' || $values['state'] === 'completed') {
+                            // Do not start the tour.
+                            continue;
+                        } elseif ($values['state'] === 'delayed') {
+                            $delay       = $params->get('delayed_time', '60');
+                            $currentTime = Date::getInstance();
+                            $loggedTime  = new Date($values['time']['date']);
+
+                            if ($loggedTime->add(new \DateInterval('PT' . $delay . 'M')) > $currentTime) {
+                                // Do not start the tour.
+                                continue;
+                            }
+                        }
```
```suggestion
                        if (empty($values)) {
                            // Do not start the tour.
                            continue;
                        }

                        if ($values['state'] === 'skipped' || $values['state'] === 'completed') {
                            // Do not start the tour.
                            continue;
                        }

                        if ($values['state'] === 'delayed') {
                            $delay       = $params->get('delayed_time', '60');
                            $currentTime = Date::getInstance();
                            $loggedTime  = new Date($values['time']['date']);

                            if ($loggedTime->add(new \DateInterval('PT' . $delay . 'M')) > $currentTime) {
                                // Do not start the tour.
                                continue;
                            }
                        }
```

Better readability?

Quy commented on administrator/language/en-GB/plg_system_guidedtours.ini:
```diff
@@ -8,7 +8,11 @@ PLG_SYSTEM_GUIDEDTOURS_BACK="Back"
 PLG_SYSTEM_GUIDEDTOURS_COULD_NOT_LOAD_THE_TOUR="Could not start the tour. No valid ID given."
 PLG_SYSTEM_GUIDEDTOURS_DESCRIPTION="System Guided Tour Plugin"
+PLG_SYSTEM_GUIDEDTOURS_HIDE_FOREVER="Hide Forever"
+PLG_SYSTEM_GUIDEDTOURS_LEARNMORE="Learn More"
```
```suggestion
```

obuisard committed:
Removed elseif

Co-authored-by: Quy <quy@nomonkeybiz.com>

obuisard committed:
Removed 'Learn more' key

Co-authored-by: Quy <quy@nomonkeybiz.com>

QuyTon committed:
Merge branch '5.2-dev' into guided-tours-autostart

HLeithner commented:
You created several event triggers, all of them need own event classes and need to be created directly.

Also I'm wondering why you call 2 events directly behind each other.

The code in the controller changing the dB should be in the model.
I only did a quick check on the phone so maybe I missed something.

Fedik commented:
Addidionaly to @HLeithner comments.
Please reduce amount of events, I mean, we always can added them later wheen needed. 
From what I see there only one event is in use `onTourRunSaveState`,

So I would suggest to keep only 2 event, but name: `onBeforeTourSaveUserState` and  `onAfterTourSaveUserState` (it more reflect what is going on). (The name `onTourRunSaveState` a bit confusing to me)

obuisard committed:
Reduce amount of events, rename of events

obuisard committed:
Changed event name for tours

obuisard committed:
Fixed typo in comment

obuisard commented:
> Addidionaly to @HLeithner comments. Please reduce amount of events, I mean, we always can added them later wheen needed. From what I see there only one event is in use `onTourRunSaveState`,
> 
> So I would suggest to keep only 2 event, but name: `onBeforeTourSaveUserState` and `onAfterTourSaveUserState` (it more reflect what is going on). (The name `onTourRunSaveState` a bit confusing to me)

I have removed the onTourRunSaveState. Now, the user action logs are created on onBeforeTourSaveUserState.
onAfterTourSaveUserState is available to later on provide feedback to user and trigger additional actions.

obuisard committed:
Removed saveTourUserState from controller for the model

obuisard committed:
Added saveTourUserState

obuisard committed:
Changed isAutostart to accept id or uid, changed saveTourUserState definition

obuisard committed:
Changed saveTourUserState call

obuisard committed:
Removed isAutoStart from saveTourUseState

obuisard committed:
Removed isAutostart from saveTourUserState

obuisard committed:
Added function's return type declarations

obuisard committed:
Added comment

obuisard committed:
Removed empty space

obuisard commented:
> The code in the controller changing the dB should be in the model.

I moved the controller's database calls to the tour model.

obuisard commented:
> You've made changes to the cypress tests here but you havent updated the joomla-cypress package

I have created https://github.com/joomla-projects/joomla-cypress/pull/27

brianteeman commented:
and then you need to update your branch to use that - ie npm update joomla-cyprus

muhme commented:
> Olivier: I have created https://github.com/joomla-projects/joomla-cypress/pull/27 
> Brian: and then you need to update your branch to use that - ie npm update joomla-cyprus

And I have created https://github.com/joomla-projects/joomla-cypress/pull/31. From my point of view we need to wait for merge and new NPM joomla-cypress release 1.1.1 to be created. After that `devDependencies` change `"joomla-cypress": "^1.1.1"` and `npm update joomla-cypress` and change back from `skipTour` to `cancelTour` is needed.

In the meantime and as https://github.com/joomla-projects/joomla-cypress/pull/27 is already merged, temporary solution is to change `package.json`:
```
     "joomla-cypress": "github:joomla-projects/joomla-cypress",
```

and run `npm update joomla-cypress`.

obuisard committed:
Update Installation.cy.js

Revert back to cancelTour as of cypress PR #31

obuisard committed:
Merge branch '5.2-dev' into guided-tours-autostart

webiedesign commented:
Test 1 - on a new instance - successful
Test 2 - opt-out of auto-starting tours -tested several times on/off all were successful
Test 3 - opt-out of auto-starting tours for a single user -tested several times on/off all were successful
Test 4 - opt-out of the welcome tour  - successful
Test 5 - run the welcome tour manually  - successful
Test 6 - auto start 2 welcome tours   - successful
Test 7 - on an updated instance - I did a fresh install of 5.1.2, the welcome tour did run once.There was no indication of the user clicking out of it with the X . The tour did not run again. My copy of the welcome tour did not copy the steps as it did in the full install. After adding a step and toggling the auto-start to yes in the tour it performed as expected- successful

obuisard commented:
> Test 1 - on a new instance - successful Test 2 - opt-out of auto-starting tours -tested several times on/off all were successful Test 3 - opt-out of auto-starting tours for a single user -tested several times on/off all were successful Test 4 - opt-out of the welcome tour - successful Test 5 - run the welcome tour manually - successful Test 6 - auto start 2 welcome tours - successful Test 7 - on an updated instance - I did a fresh install of 5.1.2, the welcome tour did run once. There was no indication of the user clicking out of it with the X . The tour did not run again. My copy of the welcome tour did not copy the steps as it did in the full install. After adding a step and toggling the auto-start to yes in the tour it performed as expected- successful

Thank you Mary.
Copying a tour does not copy the steps, unless you duplicate the tour from the list of tours.
Can you set the test to successful in the issue tracker (at https://issues.joomla.org/tracker/joomla-cms/43814)? Thank you so much!

terryburg commented:
I have tested this item :white_check_mark: successfully on dfd42fc4e28b547daea7b71ab70ee9e5e2a04e68<br /><br />Marked successful because all the desired outcomes were achieved, I think.

There were a few minor confusing items in the tests, especially Test 7:

Test 1.
"Go to Users -> Users Action Logs
Filter by the 'Guided Tours' extension. The extension will NOT be in the list if there are no recorded logs yet."
- The part about "will NOT be in" is a bit confusing at this point, since we just created a Guided Tours event in configuration 

Test 3.
"In the 'Basic settings' tab, select the option 'disallow' for the 'auto starting tours' parameter."
- the dropdown choice is "No"

Test 7.
- Kind of confusing instructions. 
- Installed 5.1.2. Initial guided tour ran. X'ed out of that, and it would not run again. At least not anytime soon.
- Installed the update to 5.2.0-alpha4. Went to dashboard and initial guided tour did not run. But it was already not running in 5.1.2. Guided Tours is no longer selected as an item to log in User Action Logs per the desired outcome. Allow Auto Starting Tours is still Yes in global configuration for Guided Tours. Unlike before when we hit "Hide Forever" and there was a log entry of the event, now there's no way for an admin to know why that tour isn't running. Other than knowing this update will disable it.   <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43814">issues.joomla.org/tracker/joomla-cms/43814</a>.</sub>

obuisard commented:
> I have tested this item ✅ successfully on [dfd42fc](https://github.com/joomla/joomla-cms/commit/dfd42fc4e28b547daea7b71ab70ee9e5e2a04e68)Marked successful because all the desired outcomes were achieved, I think.

Thank you!

> There were a few minor confusing items in the tests, especially Test 7:
> 
> Test 1. "Go to Users -> Users Action Logs Filter by the 'Guided Tours' extension. The extension will NOT be in the list if there are no recorded logs yet."
> 
> * The part about "will NOT be in" is a bit confusing at this point, since we just created a Guided Tours event in configuration

It's just to instruct that if there is no 'Guided Tour' extension in the list of extensions, it means there are no logs for it.

> Test 3. "In the 'Basic settings' tab, select the option 'disallow' for the 'auto starting tours' parameter."
> 
> * the dropdown choice is "No"

Yes, it changed recently, and I forgot to change the test description.

> 
> Test 7.
> 
> * Kind of confusing instructions.
> * Installed 5.1.2. Initial guided tour ran. X'ed out of that, and it would not run again. At least not anytime soon.

Yes, that was the previous behavior.

> * Installed the update to 5.2.0-alpha4. Went to dashboard and initial guided tour did not run. But it was already not running in 5.1.2. Guided Tours is no longer selected as an item to log in User Action Logs per the desired outcome. 

That's because no tour was run.

> Allow Auto Starting Tours is still Yes in global configuration for Guided Tours. 
> Unlike before when we hit "Hide Forever" and there was a log entry of the event, now there's no way for an admin to know why that tour isn't running. 

The Welcome tour is not running because it is not set to auto start in 5.1. It was disabled after one run. It is an update, so the Welcome tour is not meant to run on update, only on new installs.

webiedesign commented:
I have tested this item :white_check_mark: successfully on dfd42fc4e28b547daea7b71ab70ee9e5e2a04e68<br /><br />I have tested this successfully<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43814">issues.joomla.org/tracker/joomla-cms/43814</a>.</sub>

Quy commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43814">issues.joomla.org/tracker/joomla-cms/43814</a>.</sub>

QuyTon committed:
Merge branch '5.2-dev' into guided-tours-autostart

obuisard committed:
Merge branch '5.2-dev' into guided-tours-autostart

Hackwar commented:
I'm merging this now, however I'm unhappy with the additional option in the user profile. Can we change this so that this option is only added to the profile when the site actually uses the tour feature? Maybe only inject it into the form when the system plugin is enabled?

In any case: Thank you for this contribution @obuisard

Hackwar committed:
Merge branch '5.2-dev' into guided-tours-autostart

Kostelano commented:
@obuisard 

Very confusing wording. Could you please clarify what you are talking about? Perhaps rephrase it here in a simple way so that it can be translated correctly into other languages. Thank you.

> 
> COM_GUIDEDTOURS_USERSTATE_CONNECTEDONLY="Tour User state action is only for connected users."
> COM_GUIDEDTOURS_USERSTATE_STATENOTSAVED="Tour User state not saved for user %1$s tour %2$s."
> COM_GUIDEDTOURS_USERSTATE_STATESAVED="Tour User state saved for user %1$s tour %2$s."

obuisard commented:
> @obuisard
> 
> Very confusing wording. Could you please clarify what you are talking about? Perhaps rephrase it here in a simple way so that it can be translated correctly into other languages. Thank you.
> 
> > COM_GUIDEDTOURS_USERSTATE_CONNECTEDONLY="Tour User state action is only for connected users."
> > COM_GUIDEDTOURS_USERSTATE_STATENOTSAVED="Tour User state not saved for user %1$s tour %2$s."
> > COM_GUIDEDTOURS_USERSTATE_STATESAVED="Tour User state saved for user %1$s tour %2$s."

Please check PR #44093. Thank you Konstantin @Kostelano
